### ISSUE TITLE
The elba package-building process

### ISSUE THREAD

**Issue by dcao** [dcao/elba#3]
If a package is built with its required imports specified as import paths, they don't need to be specified for packages higher up the dependency chain. Combined with package aliasing, we can ensure that no module name conflicts ever occur.

We may or may not want to archive the source. Archiving the source makes it less likely that a stray ibc file will make it into the source and cause a package rebuild, but it's more complicated and less performant.

**Comment by andylokandy** [comment]
I've done some experiment based on your statements, and I'm a bit puzzled after that. So I'll try to express my understanding, figuring out what is the exact design that we want.

- My previous thoughts that the build could be fit in single ibc is wrong; we have to store all exported modules as seperated ibc files. Thus, the manifest have to support mutiple export modules, again.

- > Instead, the big idea is that we generate ipkg files for every package based on their manifest (Elba.toml) and build using Idris' ability to build from ipkg files.

  You means we could use ipkg file and idris' minimum package system as intermediate layer of building. Basically, I am against doing this for some reasons: 

  - The ipkg is planned be deprecated by core team(https://github.com/idris-lang/Idris-dev/issues/1825). And the replacement should be a well-designed manifest file and more functional package system, which means `Elba` is supposed to be a good replacement of it and take over its jobs.

  - The build from ipkg and the build from bare typecheck only differ in whether to generate `00whatever-idx.ibc`, which is only useful with global installation and repl.

  - The intermediate metadata can be completely designed from scratch, and provided by elba, if needed.

- > This is important because when you add an ordinary directory of ibc files as an import path, you have to make all of the modules it relies on available via a compiler incantation, whether by adding more import paths with -i or adding an installed package with -p. This makes that path/package globally available to all packages. 

  > If root package dcao/root depends on f/a and f/b, and each of those depends on packages with the same module name, we can't do anything about it.

  We have not been talking about using global import or `-i` `-p` in my memory. I think all of the `-i`, `-p` and --install is parts of idris' minimum package system, which should be get rid of.

If I'm right(I am quite not sure), the new build process should be:

1. Acquire a temporary build folder(`~/.elba/tmp/build/group_lightyear-0.1.0-hash/`)

2. Copy all files with `idr` extension from Source's path, keeping theirs structure.

3. Copy dependencies's build cache folder to the folders like `~/.elba/tmp/build/group_lightyear-0.1.0-hash/dep_group/dep_name/`.

3. Make compile invocations on all exported module files.

4. Copy the ibc files only of the exported modules to build cache folder(`~/.elba/build/group_lightyear-0.1.0-hash/`), keeping theirs structure.

**Comment by andylokandy** [comment]
If we take the name alias into consideration, `-i` will be not feasible. Then we come back to copying deps's files. The fix on the design above:

1. Same

2. Same

3. Copy all build cache of dependencies to the temporary build folder, with extra layer by namespace(or alias), like `~/.elba/tmp/build/andylokandy_Lightyear-0.1.0-hash/dep_group/dep_name/`.

4. Make compile invocations on all exported module files.

5. Copy the ibc files only of the exported modules to build cache folder(`~/.elba/build/andylokandy_lightyear-0.1.0-hash`), keeping theirs structure.

**Comment by dcao** [comment]
I know that the ipkg format is supposed to be deprecated and elba is supposed to take over for its duties, but building with the ipkg is the only way (as far as I know, unless we can reverse-engineer the `00whatever-idx.ibc` and create it ourselves, which is certainly a possibility) to not require all transitive dependencies to be added to the import path.

<hr />

If you want a test case of what I'm talking about:

1. In some arbitrary working directory, clone the lightyear and tomladris repositories.
2. `cd lightyear` and `idris --build lightyear.ipkg --ibcsubdir ibcs`
3. `cd ../tomladris` and `idris --build tomladris.ipkg -i ../lightyear/ibcs -p contrib --ibcsubdir ibcs`
3.1. `Tomladris.idr` should be modified such that it doesn't re-export Lightyear and all of its functions should be made `public export`, since as it is it won't compile.
4. `cd ..` and `touch Main.idr`
5. The contents of Main.idr should be the following:
```idris
module Main
import Tomladris
main : IO ()
main = do
  pure $ parseToml "irrelevant"
  pure ()
```
6. Attempt to use the normal ibc files (not generated by `idris --build`) by adding the package directory of tomladris itself as an import path: `idris -i tomladris/ -p contrib Main.idr -o main`.
6.1. Idris will error by saying that it can't import Lightyear. The only way to remedy this error would be to additionally add `-i lightyear` to the compile options. However, if we do this, now `Main.idr` would be able to also `import Lightyear`, which we don't want.
7. Instead, we use the ibc files as generated from `idris --build`: `idris -i tomladris/ibcs -p contrib Main.idr`.
7.1. Now, we can use tomladris without having to also add lightyear to our import path, and Main can no longer `import Lightyear`, as intended.

<hr />

With all that being said, I'd like to suggest a much simpler build process:
1. cd `~/.elba/src/package-hash`
2. Generate `package.ipkg` and run `idris --build -i ~/.elba/build/dependency-a-hash -i ~/.elba/build/dependency-b-hash -i <and so on for all dependencies> --ibcsubdir ~/.elba/build/package-hash` **OR** we build with `idris --check <all the files> -i <same as above> --ibcsubdir ~/.elba/build/package-hash` and elba somehow generates the `00whatever-idx.ibc` and puts it with the ibc output files.
3. Rinse and repeat.

**Comment by andylokandy** [comment]
I've sucessfully built it by following your steps, except that the compiler complained that -p contrib is not supported following `idris --build`, so I placed it in ipkg as `pkgs = contrib`.

After that, I made a small change that I deleted the `00lightyear-idx.ibc`, and rebuilt the `tomladris` and `Main`. Finally, everything worked just fine, even that `tomladris` reexported `lightyear`.

And I noticed two issues with the build process:

1. The complier stopped to redirect ibcs to `ibcsubdir` when the ibc already existed near source idr.

2. The editor plugin failed to work(the case-split, print-type and so on) because the dependencies wasn't installed at global.

**Comment by andylokandy** [comment]
What version of Idris are you using? I am using the 1.3.0 version.

The problem of plugin should be fixed at the plugin side(even compiler side), to allow users to configure extra -i path for edit-time typecheck, and it should be `/project/target/deps` by default.

**Comment by andylokandy** [comment]
Oh, I found your problem!

> 6. Attempt to use the normal ibc files (not generated by idris --build) by adding the package directory of tomladris itself as an import path: idris -i tomladris/ -p contrib Main.idr -o main.

Of course it doesn't work, since the compiler needs `lightyear` to build from source. But what I mean is to build the `tomladris` first, into `tomladris/ibcs`, and then compile Main by `-i /tomladris/ibcs`

**Comment by dcao** [comment]
Well, it turns out that I was mistaken - from my testing, it turns out that we don't even need to build with the ipkg to prevent transitive importing. So we can simplify the build process even further:

1. cd `~/.elba/src/xxx-hash`
2. `idris --check <all the source files> -i <the dependencies needed> --ibcsubdir ~/.elba/build/xxx-build-hash`
    - We only need to specify add the import paths of the package's **direct dependencies**, to avoid transitive importing
3. Rinse and repeat for all the dependencies.

Does this work for you?

**Comment by dcao** [comment]
As it turns out, I was completely mistaken on this front. If you compile a package by specifying all of the input files in `idris --check <input files>` and provide the import paths of that package's direct dependencies in the --check invocation, you don't need to provide the import paths of transitive dependencies for packages further up the chain.

With that being said, I think we can use this much simpler build process:
1. `cd ~/.elba/src/xxx-src-hash`
2. `idris --check <all of the source files> -i <all of the direct dependencies> --ibcsubdir ~/.elba/build/xxx-build-hash`
    - We only need to add the import paths of the package's direct dependencies; adding the import paths of transitive dependencies is unnecessary and would lead to the module name conflicts described previously
3. Rinse and repeat!

Does this look good to you?

**Comment by andylokandy** [comment]
No, I still think copying the direct dependencies and build in temporary folder is the only choice, in order to:

- Allow name alias, which is hard to do with -i import path.

- Avoid ibcsubdir issue. If the ibc was accidently generated in source folder(by editor extensions?), the corresponding ibc won't be generated into ibcsubdir. So there are only two solutions for us: Copy files by ourselves, or remove all ibcs in source before compilation. The latter could cause problems that the author imports ibcs in source, and we are easy to break it.

**Comment by andylokandy** [comment]
And I've realized another issue: the accidently generated ibc could change the hash of source! This behavior of compiler is so annoying... Maybe we have to zip them up in cache.

**Comment by dcao** [comment]
For name aliasing, one solution I came up with is to create a dummy package which just re-exports the package that it's aliasing, and have the compiler use the import path of that instead of the original package. In this way we don't have to copy-paste a directory and hunt for module or file names.

Using tarballs to store the Sources isn't a bad idea; it makes hashing pretty easy, and we don't have to worry about the directory of the source being mucked up by external programs. If we did go down this route, do you think we should copy-paste local dependencies into the global cache and archive them too?

Another idea I think we could consider is explicitly requiring users to specify in the manifest what files are needed for compilation, and what files are extraneous and can be deleted by elba during compilation. In effect, this does the same thing that the archiving does (except instead of implicitly specifying which files are needed by putting them in an archive, we explicitly specify that in a manifest entry); any files in the source directory that aren't listed can be deleted before compilation and ignored during hashing.

**Comment by andylokandy** [comment]
> do you think we should copy-paste local dependencies into the global cache and archive them too?

IMO, we have better not, because users is responsible for their local dependencies. Think that we have cached the local dependencies, and the user makes change in local dependency, then is the cache usable?

>Another idea I think we could consider is explicitly requiring users to specify in the manifest what files are needed for compilation ...

I think it will be too much for users; users take care of the interface and we take care to pack them up.

AFAICS, the solution of copying would not be too painful. The actual copying would only happen when:

- A dependency is built for the first time.

- Root package changed the dependencies request in manifest, so that we have to refresh /target/deps, which should be useful for edit-time typecheck.

- Root package is being build, and in this case we only copy the /target/deps to /target/build.

Additionally, if we do choose to archive the source, then we have to unpack them into a temporary folder, in which -i import does not benefit too much.

**Comment by dcao** [comment]
If we don't archive our source, stray ibc files in the source directory could still change the hash of the source, which would lead to a re-building of that package... this isn't the biggest problem in the world, it just makes caching less useful in an edge case.

<hr />

1. Get the source contents to the build directory
    - If we don't archive: `cp -r ~/.elba/src/xxx-src-hash ~/.elba/build/xxx-build-hash && cd ~/.elba/build
/xxx-build-hash`
    - If we do archive: unpack the archive into `~/.elba/build/xxx-build-hash && cd ~/.elba/build/xxx-build-hash`
2. `idris --check <all of the source files> -i <all of the direct dependencies from ~/.elba/build>`
3. Remove anything that isn't an ibc file in the build directory
4. Repeat

**Comment by andylokandy** [comment]
Agree! And I vote for archiving.

> Remove anything that isn't an ibc file in the build directory

I suggest to remove anything that isn't an ibc files, and all ibc files that are not marked as exported.

**Comment by andylokandy** [comment]
And for local deps, I want to only cache its build but not to archive its source.

**Comment by andylokandy** [comment]
I've been trying to clean up the design of cache in order to use it in build plan phase, and I think it's time to discuss about it now  since we have nailed down the detail of cache.

Here is my understanding about the cache process:

1. Run Resolver and get the Solve.

2. Cache takes the Solve and pick Source for dependencies that it thinks most fit, and then return a SourceSolve.

3. Job-plan function tries to checkout builds  correspond to the sources, from the root to bottom; add a job to JobQueue when the build cache does not exist, and stop searching when all leaves reached bottom or being cached.

4. Execute JobQueue from the bottom to the top when the dependencies with same depend-depth can be execute in parallel.

**Comment by dcao** [comment]
Nothing glaring stands out to me in the cache process you've outlined, so I think we can go ahead and start implementation. I'll start cleaning up the Cache and the retrieval process today.

**Comment by andylokandy** [comment]
I think a temp folder for global build is still necessary. Because build failure and user's cancelation would corrupt the build cache, and then make the builds depending on it fail to compile.

**Comment by andylokandy** [comment]
Do we download a tar resolution every time?

**Comment by andylokandy** [comment]
Do we download the tar resolution every time? And can we drop the support to tar resolution? 

Without tarball, Cargo only supports git and index, and it does not seem to have caused too much inconvenience.

For elba, focusing on the git allows us to iterate faster and get a cleaner implementation, and make elba run faster.

**Comment by dcao** [comment]
Good point; I'll add a temp folder for building to the Cache.

For sources, tar resolutions are only downloaded if it's not already in the cache.

**Comment by andylokandy** [comment]
The `taget/build` is supposed to be a tmp folder in #6. But now you confused me with the changes in https://github.com/dcao/elba/commit/37b9b97fcb9f222656cc5f30261d561256f07e09. Think that a user builds the lib target of his root project, and then he builds the bin target as well as the bench target and test target, and what is in `taget/build` now? Where can the user find his lib target result?

My plan is to make `target/build` a tmp folder for invocations and copy the results into corresponding folder `lib` or `bin`.

**Comment by dcao** [comment]
I intended for the same thing. Did I mess something up somewhere?

**Comment by andylokandy** [comment]
And I want to explain why I don't want the `JobQueue` to compile the root package. This is reserved for the future demand that building lib target and multiple executable in a single command, for example `elba build --bin a b c` or `elba build --example a b c`. And that is not efficient to keep checking out deps for them.

**Comment by dcao** [comment]
We can already handle this case with `Job`'s `bin_paths` field; if the `CompileMode` of the `Job` is set to `CompileMode::Bin` the code should (in the future, it isn't right now) go through the bin paths and build each one of those.

**Comment by andylokandy** [comment]
Alright, I'll see how it works at that time.

**Comment by andylokandy** [comment]
And I want to know where would you like settle the code of `compile_bin()`.

**Comment by dcao** [comment]
If you're talking about "where to deal with the issue of how to implement `compile_bin`", we can do that in a new PR (if you've already started implementation work)

**Comment by andylokandy** [comment]
I am talking about how should the current JobQueue deal with two kinds of bin target: one that can be built individually(CompileMode::Bin) and the one that should be built **with**(and after) lib target(CompileMode::Test, CompileMode::Bench).

**Comment by andylokandy** [comment]
I don't know where to start because I am familiar with neither your current strategy in JobQueue nor the cli process. 

In my plan, the JobQueue only compiles deps and let the outer functions to deal with root package, so that functions `compile_bin` and `compile_test` can have their own strategy that deciding whether to build lib target before codegen and whether to place output in `target/bin` or `target/test`. But the current implementation mixed them together.

**Comment by andylokandy** [comment]
Also, I want to make a change in cli handler: make Cache be passed in in `cli::build::build` rather than be returned from function `solve()`.

Think how `install` subcommand work:
 
- First, we get the package, no matter from remote or loacl, and we place the package from remotein in a tmp folder.

- Then, we invoke `elba build` on it. Of course, we invoke the command internally, by reusing the build command handler function.

- Finnaly, when the build finished, we move the built executable into global bin/ folder.

As you can see, the Cache is involved in the whole process, to retrieve remote package, to build it, and to store it. So we should reuse Cache by making it a argument of the second step(or to extract some of it's code into a new function).

**Comment by andylokandy** [comment]
Should we allow `elba build` and `elba build --bin` for user to decide which target should be build? As well as `elba install` for global library and `elba install --bin` for global executable?

**Comment by dcao** [comment]
I think we should allow the user to specify `--bin` or `--lib` flags or whatever depending on which targets they want to have built or installed.

I wanted to make the JobQueue build the root package along with everything else because it seems like building the root package shares a lot of commonalities with building any other package, with the only difference being the different possible CompileModes (this flexibility helps for e.g. if we want to add task support at some point, since those will have to generate binaries too during the build process). The build system is kinda incomplete at the moment so it is pretty confusing, and you raised an excellent point in that some generated binaries have to depend on the library itself, so I had the following hypothetical system in mind:

  - **Add a new CompileMode for binaries which depend on its own package**

    This addresses your concern with binaries that need to depend on their own package's libraries. This would subsume tests, benchmarks, and possibly examples. This CompileMode might have a field which determines which folder the output executable should go in. One open question at this point would be how we could have tests and benchmarks for packages which don't export a library at all. Would creating tests/examples just be impossible for packages without libraries? And should we be able to test private, non-exported modules? 

  - During the build process, **use JobQueue to build every target that needs to be built**

    This means that JobQueue is responsible for building library output, tests, benches, and maybe even docs. While the JobQueue will automatically store built libraries into the global cache under `~/.elba/build`, for all other artifacts it shouldn't do anything, it should just keep them in the tmp dirs (either the global one or the local `target` one). JobQueue can be already be passed a parameter `root_ol` which tells the JobQueue where to store the output of the root package if not in its usual location under `~/.elba/tmp/<hash>`. The JobQueue shouldn't handle storing anything of the root package (right now it moves the root package build stuff to the global cache, which it shouldn't do; that's pretty easy to fix though). `JobQueue::exec` should also probably somehow return the OutputLayout (or the path to it) of the root package, in case the root package was global.

  - **Each subcommand is responsible for dealing with the outputs of the root package.**

    For a command like `build`, we don't have to do anything after `JobQueue::exec` is done, since all the relevant output will already be in the target folder. For something like `install`, we would take all the outputs in the `bin` folder under the OutputLayout and move them into the global `bin` directory, errorirng if we were to overwrite something. For `test` or `bench`, we would just execute all the generated test/bench executables, etc.

**Comment by dcao** [comment]
I've implemented most of the design I've described above (save for the actual compilation part) for the JobQueue in [the most recent commit](https://github.com/dcao/elba/commit/bfc4829d337bf1e7e8e8e0f8b44e04c9b91d226d). `job.rs` also now contains a small discussion of the tradeoffs between having a single, monolithic JobQueue responsible for building everything and your design: a separation between dependency preparation (handled exclusively by the JobQueue) and target generation (handled by something else). Hopefully with this more fully-featured code you can see where I'm coming from with my design.

**Comment by andylokandy** [comment]
Sounds great to me. I don't have any concern about your design. (・∀・)

Just bringing up some issues to take into acount:

- Test notation: idris don't provide a test notation like `#[test]` in rust. The current approach in ipkg system is to declear every test function in manifest. We have to figure out whether we stick with this approach or we use test notation like `%test`, which requires us to submit PRs to idris compiler.

- Codegen backend: different backends produce different executable formats, then how do we run them? How do we test or even benchmark them?(Think php and js).

- Benchmark framwork: Idris' ecosystem lacks a framework for benchmarking, so should we postpond the bench command in elba until somebody(or we) makes one.

**Comment by dcao** [comment]
- One solution I can think of for dealing with tests is to require that users create separate files for their tests, and then when building those test executables, we add the `output_layout.lib` path as an import path.

    - The main deficiency with this design is that we can't test non-exported modules; there's probably a way we can work around that somehow

- If a user changes compilation backends (presumably through some sort of configuration change), maybe they should be able to specify some way of running the executables (all js output should be run with `./node`, php output should be with `./php`, etc.).

- I think knocking off the bench command makes sense.

**Comment by andylokandy** [comment]
- Test

I am not worry about how we build tests but worry about how we run tests; testing turns out to be not just a bin target now. You can read this to get my point of "test notation" http://docs.idris-lang.org/en/v0.12/tutorial/testing.html.

- Backend 

Maybe we add a backend section to manifest? For example:

[[backend]]
c = ""
js = "node"
default = "js"

// or
[backend.js]
executor = "node"

**Comment by andylokandy** [comment]
Maybe postpond test as well?

**Comment by dcao** [comment]
For specifying backends, I think that belongs in the configuration file rather than the manifest file - package users should be the ones controlling backend options (for instance, my node executable could be named `zzzsdjf`; the package author can't know that).

**Comment by andylokandy** [comment]
Yeah, you're right. The backend config belongs to the end user.

And how about to only support c backend for v0.1.0?

**Comment by dcao** [comment]
I don't think supporting other backends for simple binary generation would be too bad (it should just be a command line argument plus the configuration file updates if we want an `elba run`).

Testing, on the other hand... I'm not so sure.

**Comment by andylokandy** [comment]
I made a test example just now, and it showed me that we can just build tests as common executable target.

```idris
module Main

import Test.Unit

testA : IO Bool
testA = assertEquals (1 + 1) 2

testB : IO Bool
testB = assertTrue (1 + 1 == 2)

main : IO ()
main = NonReporting.runTests [testA, testB]
```

```bash
> idris -o main .\main.idr --package contrib
> .\main.exe
Test: Assert Equals
Test: Assert True
All Tests have been performed.
========================================
```
